[Liu Chen En] iP

twothicc opened the pull request:
# Duke
> "MMMM so good and tasty" - Merril Howard Kalin ([source](https://www.youtube.com/watch?v=-FrpuPLYnvY))

If you aren't using Duke, then why are you even **ALIVE**? It's,

- text-based.
- easy-to-learn.
- ~~FAST~~ ULTRA FAST to use.

See how an average non-Duke user compares to the average Duke user.
![Average non-Duke user vs Average Duke enjoyer](https://i.imgflip.com/556g88.jpg)

Stoked? Wait no more!
All you need to do is,

1. Download it from [here](https://github.com/twothicc/ip/releases/tag/v0.1).
2. Double-click it.
3. Add your tasks.
4. Let *ol* Duke take care of managing your tasks. :100:

**OH** did I mention that Duke is absolutely **FREE**?

That's right. We don't need you to sell your *soul* for our application. No one should have to do so to manage their tasks.

Features:

- [x] Storing general tasks and tasks with a certain date.
- [x] Marking tasks as done.
- [x] Deleting tasks.
- [x] Finding tasks by date and keyword.
- [ ] Reminders  (coming soon)

* * *

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public static void main(String[] args) {
        Duke duke = new Duke(SAVE_FILENAME);
        duke.run();
}
```

Commit:
Add Gradle support

twothicc committed:
Added Greet, Echo, Exit functionality to chatbot with formatting

twothicc committed:
Add the chatbot ability to store text input from the user and display them back to the user when requested

twothicc committed:
added Task class

twothicc committed:
Add chatbot functionality that allows users to mark tasks as done

twothicc committed:
Added Deadline class

twothicc committed:
Added Todo class and overrode toString() in Task class

twothicc committed:
Added Event class

twothicc committed:
Added support for tracking three types of tasks

twothicc committed:
Updated input.txt and EXPECTED.TXT for automated text ui testing

twothicc committed:
Added DukeException class

twothicc committed:
Added error handling for various incorrect user inputs

twothicc committed:
Updated automated ui testing to test error handling too

twothicc committed:
Changed storage for user tasks from Array to ArrayList

twothicc committed:
Added chatbot functionality to delete stored tasks

twothicc committed:
Added enums Commands to replace commonly used chatbox commands

twothicc committed:
Changed chatbot to close scanner after bye and removed some redundant function calls

twothicc committed:
Created test directory to store JUnit test classes and installed JUnit 5 dependency. Created dummy JUnit tests for Task class in TaskTest class.

twothicc committed:
Added unit tests for public members of Task class in TaskTest class

twothicc committed:
Improved code readability by creating new enums and abstracting some functions. Also updated EXPECTED.TXT to fit new changes.

twothicc committed:
Added methods to read and write to /data/dukeSave.txt file. Added /data/ to .gitignore so save files are not committed.

twothicc committed:
Implemented save and read functionality. Changes made to Task, Todo, Deadline and Event classes to facilitate converting Tasks to save format. .gitignore updated to ignore save files in text-ui-test.

twothicc committed:
Added functionality to parse LocalDate into formatted strings and to convert formatted strings into LocalDate. Deadline and Event objects now store LocalDate instead of any string. Added 2 new DukeException messages that corresponds to failure to parse LocalDate and failure to convert to LocalDate.

twothicc committed:
Fixed bug where empty task description is allowed for Event and Deadline in addTask() method.

twothicc committed:
Merge branch 'branch-Level-7'

twothicc committed:
Merge branch 'branch-Level-8'. Resolved conflicts in Duke.java. Added toSaveDateFormat to convert LocalDate into a string suitable for saving. Edited toSaveFormat and Deadline, Event and Task classes in response to changing by and at to LocalDate object.

twothicc committed:
Added Parser class. Organized methods to do with parsing user commands and inputs into Parser class.

twothicc committed:
Added Storage class. Organized methods and variables that deal with loading and saving into Storage class.

twothicc committed:
Added TaskList class. Organized methods that involve operations on tasks into TaskList.

twothicc committed:
Add Ui class

Enums are clogging up space in Duke.

As commands and other recognized inputs by Duke chatbot increases later, it can get increasingly difficult to navigate around and manage code.

Enums are extracted to Ui class.

Change to have Duke constructor initialize storage and tasklist objects

Storage class methods are static and Duke does not have a constructor.
Multiple methods are being called in main method, which is bad for code readability.
Filename is kept as a static constant in Storage class and can be difficult for developers to find later on.

Abstract Storage and Tasklist initializations away to Duke class constructor.
Move scanning for and handling user inputs into run method in Duke class.
Move Filename for save file to Duke where main method is at.

Developers set up Duke chatbot by just creating a Duke object and can run it by calling run method.
Developers can easily see the filename for the save file in Duke class.

twothicc committed:
Add methods to Ui class to print standard responses

twothicc committed:
Change readTasksFromData and saveTasksToData methods in Storage to take TaskList objects as parameters

twothicc committed:
Replace println calls to print responses with Ui methods that achieve the same purpose

twothicc committed:
Extracted out code into classes extending from an abstract Command class

twothicc committed:
Extracted out exception messages to be returned by methods in Ui class

twothicc committed:
Organize classes into packages duke, duke.command and duke.task

twothicc committed:
Add JUnit5 tests for public methods of classes in duke.task package

twothicc committed:
Create Storable interface and add StorageStub class

No Storable interface.

Unable to create Stub for Storage class.

Add Storable interface and have StorageStub and Storage class implement it.

StorageStub class can be created for unit testing purposes.

Add JUnit5 tests for AddCommand and introduced new exception messages

Exception messages do not check for missing spaces in user input.

Added new checks and appropriate exception messages.

This is so user can better understand why their input is invalid.

twothicc committed:
Add unit test with equals method to Task, Todo, Event and Deadline classes

twothicc committed:
Add checks to handle invalid date related inputs by user

twothicc committed:
Add JUnit5 tests for DeleteCommand. Extract preliminary input checking to method in Parser

twothicc committed:
Add JUnit5 tests for MarkCommand class. Organize JUnit5 tests for DeleteCommand

twothicc committed:
Add ParserTest class for JUnit5 tests. Add equals method to Command classes with corresponding unit tests

twothicc committed:
Add JUnit5 tests to ParserTest class. Add new exception messages to Ui class

twothicc committed:
Optimize package imports. Add JUnit5 tests for TaskList in TaskListTest class

twothicc committed:
Add manifest.mf file for building artifact

twothicc committed:
Add Javadocs for classes in duke.command package

twothicc committed:
Add JavaDocs for classes in duke.task package

twothicc committed:
Add JavaDocs for packages in duke

twothicc committed:
Change to fit coding standard

twothicc committed:
Add functionality to search for tasks by a keyword

twothicc committed:
Merge branch 'branch-Level-9'

twothicc committed:
Merge branch 'branch-A-CodingStandard'

twothicc committed:
Merge branch 'branch-A-JavaDoc'. Resolve merge conflicts in Ui class

twothicc committed:
Add JavaDocs to FindCommand class and methods in Ui class that generate standard responses for Find operation

twothicc committed:
Improve code to fit Coding Standard

twothicc committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

twothicc committed:
Setup build.gradle to build Duke.jar using Gradle. Move test files into src/test/java/ to allow unit testing using Gradle

twothicc committed:
Merge branch 'branch-A-Gradle'

twothicc committed:
Setup Checkstyle to match that of AB3 and change package import order

twothicc committed:
Resolve Coding Standard violations found by Checkstyle

twothicc committed:
Merge branch 'branch-A-CheckStyle'

twothicc committed:
Set up basic GUI for Duke using javafxml

twothicc committed:
Change Duke GUI

twothicc committed:
Merge branch 'branch-Level-10'

twothicc committed:
Remove unnecessary code for running text UI based Duke

twothicc committed:
Fix found task not being converted to String

leezhixuan commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,73 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
+    public static DialogBox getDukeDialog(String text, Image img) {
+        DialogBox db = new DialogBox(text, img);
+        db.flip();
+        return db;
+    }
+
+    public static DialogBox getDukeWelcome(Image img) {
+        DialogBox db = new DialogBox("Duke: Hello! I'm Duke\nWhat can I do for you?", img);
+        db.flip();
+        return db;
+    }
+
+    public static DialogBox getDukeBye(Image img) {
+        DialogBox db = new DialogBox("Duke: Bye. Hope to see you again soon!", img);
+        db.flip();
+        return db;
+    }
```
Did you forget to add javadocs for these public methods?

twothicc replied:
Thank you for pointing the missing javadocs out to me!
I'll add it in in the next push.

leezhixuan commented on src/main/java/duke/DukeLauncher.java:
```diff
@@ -0,0 +1,9 @@
+
+import javafx.application.Application;
+
+public class DukeLauncher {
+    public static void main(String[] args) {
+        Application.launch(Main.class, args);
+    }
```
Are javadocs missing for this class and its public methods?

twothicc replied:
It was a very recent addition. I'll add javadocs for the class and its public methods in the next push.

leezhixuan reviewed (commented):
I think your work is incredibly neat and well documented. Not so sure if it matters but I thought that it might be good to add in some descriptions for your test cases. Good work!

eezj35 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+ */
+public class Duke {
+    /** Variables related to initializing Duke */
+    private static final String SAVE_FILENAME = "dukeSave.txt";
```
I like that you cached the filename as a constant variable!

eezj35 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+        this.storage = new Storage(fileName);
+        this.ui = new Ui();
+
+        // Read tasks from save file.
```
I like the comments, very helpful in making the code more readable 👍

eezj35 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,274 @@
+
+    /**
+     * Parses a LocalDate into a string format "d MMMM yyyy" to be printed to the user.
+     * For instance, the LocalDate equivalent of 2020-01-01 will be parsed into "1 January 2020".
```
I like the detailed javadocs summary!

eezj35 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,274 @@
+    public static LocalDate toLocalDate(String dateString) throws DukeException {
+        String[] split = dateString.split("[/\\s]");
+
+        // Check if split has 3 elements. If not, this is already an invalid date format.
```
Perhaps a comment here might no be as necessary? I think the code is simple enough!

twothicc replied:
I agree that the code is simple, but I would still prefer to keep the comment for clarity purposes.

eezj35 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,274 @@
+            return new ListCommand(userInput);
+        } else {
+            if (userInput.startsWith(Commands.DONE.getCommand())) {
+                // Mark task as done.
```
Perhaps a comment might not be so necessary here? I think the way you have written your code is quite understandeable without comments!

twothicc replied:
Thanks for pointing this out! This comment does seem to be stating the obvious and could instead contribute to clutter in the code. I will remove it in the next push.

eezj35 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,274 @@
+                // Mark task as done.
+                return new MarkCommand(userInput);
+            } else if (userInput.startsWith(Commands.DELETE.getCommand())) {
+                // Delete a task.
```
Same comment here!

eezj35 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,274 @@
+            // Add 1 as user's number input is separated from command by 1 space.
+            return Integer.parseInt(userInput.substring(command.getLength() + 1));
+        } catch (NumberFormatException nfe) {
+            // Invalid user input cannot be parsed into Integer.
```
Noticed the same issue about comments here! Maybe you could check your other methods to see if the comments are necessary?

eezj35 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,243 @@
+    private static final String SAVE_SEPARATOR = " ~ ";
+
+    /** Filename of file to be saved to */
+    private final String saveFileName;
```
This variable sounds a little like a verb, maybe you could change it to be a noun?

twothicc replied:
Thanks for the suggestion! However since "separator" is a noun and that it sufficiently describes the purpose of this constant, I do not feel that it is necessary to make changes to this name.

eezj35 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,243 @@
+     */
+    private static Task parseSaveFormat(String save) throws DukeException {
+        // Split save string by the save separator.
+        String[] saveSplit = save.split(SAVE_SEPARATOR);
```
Perhaps this variable name could be made into a noun too?

twothicc replied:
Thanks for the suggestion! I agree that it would be more appropriate to use a noun as the name of this variable and I will make the changes in the next push.

eezj35 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,94 @@
+     * @return The removed Task.
+     */
+    public Task remove(int idx) {
+        Task tempTask = this.tasks.get(idx);
```
Maybe could rename this variable to taskToBeRemoved?

twothicc replied:
Thanks for the suggestion! That is indeed a more descriptive name. I will make the changes in the next push.

eezj35 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,364 @@
+    }
+
+    /** Enums for Duke chatbot commands */
+    public enum Commands {
```
I like your use of Enums here

eezj35 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,364 @@
+     *
+     * @return Standard goodbye response.
+     */
+    public String getGoodbye() {
```
Maybe for this method (and similar methods in this class), you could rename the methods to this format: 
"<command>Message". So for this case, maybe something like "goodbyeMessage". Feel free to use or not use my suggesstion though!

twothicc replied:
Thanks for the suggestion! I agree that this will provide a better description for the function's purpose and I will make relevant changes in the next push.

eezj35 reviewed (commented):
Hi Great job on your code! I think your methods are all of suitable length and your code is generally quite readable, although i think some comments can be removed if the line of code it is referencing is easy to understand. Taking away some comments can reduce the clutter too! I like your OOP approach as well. Overall really good job!

twothicc committed:
Add changes suggested by PR reviews

twothicc committed:
Change main class  for Duke application

twothicc committed:
Add assertions to validate assumptions in code

Assumptions are that users will not pass null as a String argument to Task-related
constructors or to parseSaveFormat method.
Similarly, assumptions are also made that users will not pass empty string or null into DialogBox constructor
and that they will also not pass a null as Image argument into DialogBox constructor.
Assertions need to be made to ensure that later users do not misuse these methods/constructors.

Assertions are added to corresponding constructors/methods to validate the aforementioned assumptions.

twothicc committed:
Remove misuse of assert statements and add new assert statements for DialogBox flip method

Assert statements were used to validate that arguments passed into public methods were not null,
when it is more appropriate to let it raise runtime exceptions like NullPointerException that describe
the error better.

Remove misused assert statements and add assert statements for flip method of DialogBox.

Flip method of DialogBox assumes that the HBox is already aligned to the top right of its container.

twothicc committed:
Remove redundant comments

Redundant comments used during implementation are left in the methods. These comments
are unnecessary as the code is simple enough to be self-explanatory.

Redundant comments are removed.

Code appears to be less cluttered and readability is improved.

twothicc committed:
Combine Ui methods that are unnecessarily split in two

Methods in Ui return parts of the same standard response and can be combined for better readability.

Methods in Ui that return parts of the same standard response are combined.

Only one method for each standard response makes code more readable and easier to use.

twothicc committed:
Rename Commands enum class to UserCommands to differentiate from Command class

twothicc committed:
Split DukeException into more subclasses to represent exceptions specifically for better readability

twothicc committed:
Change to read from save file using streams for improved performance on large files

twothicc committed:
Merge pull request #1 from twothicc/branch-A-Assertions

Branch a assertions

twothicc committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/MainWindow.java
#	src/main/java/duke/command/DateCommand.java

twothicc committed:
Merge branch 'master' into branch-A-Streams

twothicc committed:
Merge pull request #2 from twothicc/branch-A-CodeQuality

Branch a code quality

twothicc committed:
Merge branch 'master' into branch-A-Streams

twothicc committed:
Merge pull request #3 from twothicc/branch-A-Streams

Branch a streams

twothicc committed:
Setup CI using Github Workflow

twothicc committed:
Resolve package import checkstyle errors

twothicc committed:
Change Task class to abstract and remove TaskTest class

twothicc committed:
Group Deadline and Event classes under abstract class TimedTask

twothicc committed:
Add PeriodTask class

twothicc committed:
Add functionality to allow Duke to recognize a Period task

twothicc committed:
Add logic to handle when user provides a chronologically invalid time period

twothicc committed:
Add functionality to Duke to allow it to find Period tasks by date

twothicc committed:
Merge branch 'branch-B-DoWithinPeriodTasks'

twothicc committed:
Create DukeResponse to classify responses as error and non-error messages

twothicc committed:
Add visual changes to GUI

twothicc committed:
Apply SLAP to long methods and remove unused MissingDescriptorException

twothicc committed:
Change naming for variables used in test cases to be more descriptive

twothicc committed:
Fix issues with background image moving with scroll

twothicc committed:
Change index shown for matching task to be its actual index in tasklist

twothicc committed:
Add brief user guide and screenshot showcase

twothicc committed:
Fix wrong anchor link in user guide

twothicc committed:
Fix index shown to be 1-based

twothicc committed:
Fix issue involving skipping in loop

twothicc committed:
Remove unnecessary print statements

twothicc committed:
Remove unnecessary print statements

twothicc committed:
Fix long methods

twothicc committed:
Remove dead code
